### PR TITLE
GCP Mailer Provider functionality

### PR DESCRIPTION
**Pull request by bclements**
This enhancement to the c7n_mailer tool adds gcp support. The following items were included in this PR:

- GCP Pub Sub message processing.
- Notification delivery via EmailDelivery. (Slack & Datadog integration parity coming soon in subsequent PRs)
- Basic Unit Tests
- Modification/updates to README.md to include GCP Config information.

Functionality has been tested and verified using mailtrap.io smtp testing servers. 

The process flow for GCP Mailer is as follows:

- Custodian GCP Policy sends notification to GCP PubSub Topic as normal.
- C7N_Mailer is run from the command line (Google Cloud functions are not supported at this time)
- The GCP Mailer processor retrieves all messages from the queue and processes each message.
- For each message, it will send notification to the list of "To" email addresses using SMTP infrastructure configured in the config file.

How to test this code:

1. gcloud auth login
2. Create a gcp pubsub topic(for instance: gcp-notify-test1) and add a "PULL" subscription to that topic
3. Create an example policy similar to this or use another gcp resource which as resources:

```yaml
policies:
  name: gcp-notify-test-for-mailer
  resource: gcp.function
  actions:
    type: notify
      template: default
      subject: 'testing the c7n mailer'
      to:
        testuser@example.com
      transport:
        type: pubsub
        topic: gcp-notify-test1
```
4. Run python cli.py -c mailerconfig.yml --run

### CODE REVIEW

File: tools/c7n_mailer/c7n_mailer/cli.py [hunk 1]
```diff
@@ -10,8 +10,10 @@
 from c7n_mailer import deploy, utils
 from c7n_mailer.azure.azure_queue_processor import MailerAzureQueueProcessor
 from c7n_mailer.azure import deploy as azure_deploy
+from c7n_mailer.gcp.gcp_pubsub_processor import MailerGcpPubSubProcessor
 from c7n_mailer.sqs_queue_processor import MailerSqsQueueProcessor
-from ruamel import yaml
+from ruamel import yml
```
Comment (thisisshi) [inline tools/c7n_mailer/c7n_mailer/cli.py]: this import doesn't look right, looks like you did a find and replace on yaml/yml

File: tools/c7n_mailer/c7n_mailer/cli.py [hunk 2]
```diff
@@ -207,6 +213,9 @@ def main():
 
         if is_azure_cloud(mailer_config):
             azure_deploy.provision(mailer_config)
+        elif is_gcp_cloud(mailer_config):
+            print('Deploying mailer as a google cloud function is not supported at this time.')
```
Comment (thisisshi) [inline tools/c7n_mailer/c7n_mailer/cli.py:217]: raise something here so it's not just lost in the logs

File: tools/c7n_mailer/c7n_mailer/email_delivery.py [hunk 3]
```diff
@@ -74,7 +74,8 @@ def __init__(self, config, session, logger):
         self.config = config
         self.logger = logger
         self.session = session
-        self.aws_ses = session.client('ses', region_name=config.get('ses_region'))
+        if not self.config.get('queue_url').startswith('projects'):
```
Comment (thisisshi) [inline tools/c7n_mailer/c7n_mailer/email_delivery.py]: reuse the is_gcp_cloud function you defined in cli

File: tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                'max_messages': MAX_MESSAGES
+            }
+        })
+
+    def ack_messages(self, discard_datetime):
+        """Acknowledge and Discard messages up to datetime using seek api command
+        """
+        client = self.session.client('pubsub', 'v1', 'projects.subscriptions')
+
+        return client.execute_command('seek', {
```
Comment (thisisshi) [inline tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py]: why use seek? ack is more appropriate here?

File: tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+MAX_MESSAGES = 1000
+
+
+class MailerGcpPubSubProcessor(object):
+
+    def __init__(self, config, logger, session=None):
+        self.config = config
+        self.logger = logger
+        self.subscription = self.config['queue_url']
+        self.session = session or Session()
```
Comment (thisisshi) [inline tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py:31]: set up the client here as it's reused elsewhere in the other functions

File: tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    def run(self):
+        self.logger.info("Downloading messages from the GCP PubSub Subscription.")
+
+        # Get first set of messages to process
+        messages = self.receive_messages()
+
+        while len(messages) > 0:
+            # Discard_date is the timestamp of the last published message in the messages list
+            # and will be the date we need to seek to when we ack_messages
+            discard_date = messages['receivedMessages'][-1]['message']['publishTime']
```
Comment (thisisshi) [inline tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py:43]: properly acking is preferred here imo this is kind of hacky, you can provide a list of ack ids so you can ack a bunch of messages at once

File: tools/c7n_mailer/c7n_mailer/cli.py (hunk 1 above)
Comment (bclements) [reply tools/c7n_mailer/c7n_mailer/cli.py]: Yeah it was a typo on my part. I accidentally removed that import and typed it back in wrong. Fixed in next commit.

File: tools/c7n_mailer/c7n_mailer/cli.py (hunk 2 above)
Comment (bclements) [reply tools/c7n_mailer/c7n_mailer/cli.py:217]: good advice. Fixed by adding in raise Warning in next commit.

File: tools/c7n_mailer/c7n_mailer/email_delivery.py (hunk 3 above)
Comment (bclements) [reply tools/c7n_mailer/c7n_mailer/email_delivery.py]: Good catch. Fixed in the next commit.

File: tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py (hunk 6 above)
Comment (bclements) [reply tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py:43]: Yeah this is actually kind of hacky. I struggled with this earlier in the year until I finally figured out how to handle this.  Pubsub is eventually consistent so what would happen is that if you ack'd a list of message id's...they would still be around the next time you retrieved messages. I'd have to keep an every growing list of acknowledged id's. Didn't seem too clean to me. Reading into Seek, it seems that is the recommended way of handling this use case using discard dates, etc. I'll have to find the article I read that convinced me of this. I'm going to upstream commit it like this and if push comes to shove I'll have to proof what I'm talking about to you guys.

File: tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py (hunk 5 above)
Comment (bclements) [reply tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py:31]: Fixed in next commit.

**Pull request closed without merging**